### ISSUE TITLE
Weights & Biases initialization should be decoupled from the Logger

### ISSUE THREAD

**Issue by soumik12345** [Deci-AI/super-gradients#980]
Currently, a Weights & Biases run is initialized inside the constructor of [`WandBSGLogger`](https://github.com/Deci-AI/super-gradients/blob/master/src/super_gradients/common/sg_loggers/wandb_sg_logger.py#L100). However, if the run initialization is not decoupled from the logger, it would not be possible for a user to use [Weights & Biases Artifacts](https://docs.wandb.ai/guides/artifacts) for storage and versioning of their datasets. Hence, `wandb.init()` should be **optionally** decoupled from `WandBSGLogger`.

The ideal workflow should be:

```python

wandb.init(project=..., entity=..., **kwargs)

artifact = wandb.use_artifact('geekyrakshit/yolo-nas-integration/bdd100k-subset-yolo:v2', type='dataset')
artifact_dir = artifact.download()

train_data, val_data = create_dataset(artifact_dir)
model = ...

trainer = Trainer(experiment_name=..., ckpt_root_dir=...)
trainer.train(model=model, training_params=train_params, train_loader=train_data, valid_loader=val_data)
```

**Comment by BloodAxe** [comment]
I see your point, however it is not quite how we thought users would use SG ^_^
We build SG to rely heavily on use of recipes, so for that purpose initialization of logger inside trainer made perfect sense. 

What we can do is a small change inside the [WandBSGLogger](https://github.com/Deci-AI/super-gradients/blob/master/src/super_gradients/common/sg_loggers/wandb_sg_logger.py#L100) to check whether there is active run (`wandb.run is not None`) and skip call to `wandb.init(...)` if it was already made. 

Although it looks simple from the first glance, I feel there could be edge cases with resume (?) @shaydeci @Louis-Dupont what you guys think about it?

**Comment by soumik12345** [comment]
> WandBSGLogger

Hi @BloodAxe 
I think this could be resolved by changing [this line](https://github.com/Deci-AI/super-gradients/blob/master/src/super_gradients/common/sg_loggers/wandb_sg_logger.py#L100) to

```python
run = wandb.init(
    project=project_name, name=experiment_name, entity=entity, resume=resumed, id=wandb_id, **kwargs
) if  wandb.run is None else wandb.run
```

This would mean that if a run is already initialized, `WandBSGLogger` won't attempt to initialize a run again, else it would be automatically initialized as per the default behaviour. Would love to know your thoughts on this.

**Comment by ofrimasad** [comment]
@soumik12345 I am afraid it is a bit more complicated...
For example, assume that you want to resume a training run, how can the logger know that the run is the correct one? what if it is not? 
In your initialization you are not passing  `name`, where will the metrics be save? to which experiment?

**Comment by soumik12345** [comment]
> For example, assume that you want to resume a training run, how can the logger know that the run is the correct one? what if it is not?

Hi @ofrimasad 
In order to resume a run, you can do it without using `wandb.save()` or `wandb.restor()` by initializing your run using

```python
wandb.init(id="id_of_run_to_be_resumed", resume="must")
```
If this is used, `wandb` will start logging in that same run from the last step.

> In your initialization you are not passing name, where will the metrics be save? to which experiment?

If you use the aforementioned method to resume your run, then the metrics will be saved in the same run/experiment, logged from the last step where it was left to.

**Comment by kldarek** [comment]
@ofrimasad @BloodAxe - the pattern suggested by @soumik12345 is something we've implemented in numerous integrations, including e.g. Pytorch Lightning, Keras etc. and it's been working well for the users, we haven't really seen issues related to resuming runs. How could we move forward on this or discuss any blockers?
